File: src/main/java/data/priorityQueue/ReorganizeString_767.java
package data.priorityQueue;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.PriorityQueue;

public class ReorganizeString_767 {
  public static String reorganizeString(String s) {
    var count = new HashMap<Character, Integer>();
    for (var c : s.toCharArray()) {
      count.put(c, count.getOrDefault(c, 0) + 1);
    }

    var pq = new PriorityQueue<Character>((a, b) -> count.get(b) - count.get(a));
    for (var k : count.keySet()) {
      pq.offer(k);
    }

    var reorg = "";
    while (!pq.isEmpty()) {
      if (pq.size() == 1 && count.get(pq.peek()) > 1) {
        return "";
      }
      // pop 2 chars
      var group = new ArrayList<Character>();
      for (var i = 0; i < 2; i++) {
        if (!pq.isEmpty()) {
          var t = pq.poll();
          reorg += t;
          count.put(t, count.get(t) - 1);
          if (count.get(t) > 0) {
            group.add(t);
          }
        }
      }
      group.forEach(pq::offer);
    }

    return reorg;
  }

  public static void main(String[] args) {
    System.out.println(reorganizeString("aab"));
    System.out.println(reorganizeString("aaab"));
  }
}

File: src/main/kotlin/data/structure/priorityQueue/ReorganizeString_767.kt
package data.structure.priorityQueue

import java.util.PriorityQueue

class KtReorganizeString_767 {
    fun reorganizeString(s: String): String {
        val count = IntArray(26)
        val pq = PriorityQueue<Char> { a, b -> count[b - 'a'] - count[a - 'a'] }

        s.forEach { count[it - 'a']++ }
        count.indices.filter { count[it]> 0 }.forEach { pq.offer(('a'.plus(it))) }

        var result = ""
        while (pq.isNotEmpty()) {
            val chars = ArrayList<Char>()
            for (i in 0..1) {
                if (pq.isEmpty()) {
                    // not enough chars
                    break
                }
                val ch = pq.poll()
                // now we have adjacent same chars
                if (result != "" && ch == result.last()) {
                    return ""
                }
                result += ch
                count[ch - 'a']--
                if (count[ch - 'a'] > 0) {
                    chars.add(ch)
                }
            }
            chars.forEach { pq.offer(it) }
        }

        return result
    }
}

fun main() {
    val reorganize = KtReorganizeString_767()
    println(reorganize.reorganizeString("aab"))
    println(reorganize.reorganizeString("aaab"))
}